File: ios/Classes/plugin/BlueThermalPlusPlugin.swift
import Flutter
import UIKit

public final class BlueThermalPlusPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

  private var eventSink: FlutterEventSink?
  private let store = DeviceStore()
  private lazy var ble = ZebraBLETransport(store: store)
  private lazy var classic = EAClassicTransport(store: store)
  private lazy var epson = EpsonEposTransport()
  private lazy var router = TransportRouter(ble: ble, classic: classic, epson: epson)
  private var bleConfig: [String: Any] = [:]
  private var classicConfig: [String: Any] = [:]
  private var epsonConfig: [String: Any] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = BlueThermalPlusPlugin()

    let method = FlutterMethodChannel(
        name: "blue_thermal_plus/methods",
        binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: method)

    let events = FlutterEventChannel(
        name: "blue_thermal_plus/events",
        binaryMessenger: registrar.messenger()
    )
    events.setStreamHandler(instance)
  }

  // MARK: - EventChannel
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.eventSink = events
    // Redireciona eventos dos transportes para o Flutter
    ble.onEvent = { [weak self] payload in self?.eventSink?(payload) }
    classic.onEvent = { [weak self] payload in self?.eventSink?(payload) }
    epson.onEvent = { [weak self] payload in self?.eventSink?(payload) }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    ble.onEvent = nil
    classic.onEvent = nil
    epson.onEvent = nil
    return nil
  }

  // MARK: - MethodChannel
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]

    // Identifica qual transporte usar (default: ble)
    let transportStr = (args["transport"] as? String) ?? "ble"
    let transport = TransportType(rawValue: transportStr) ?? .ble
    let manager = router.manager(for: transport)

    switch call.method {

        // ✅ CONFIGURE: Salva e aplica as configurações globais
    case "configure":
      if let b = args["ble"] as? [String: Any] {
        bleConfig = b
        applyBleConfigFrom(b)
      }
      if let c = args["classic"] as? [String: Any] {
        classicConfig = c
        applyClassicConfigFrom(c)
      }
      if let e = args["epson"] as? [String: Any] {
        epsonConfig = e
        applyEpsonConfigFrom(e)
      }
      result(nil)

    case "startScan":
      manager.startScan()
      result(nil)

    case "stopScan":
      manager.stopScan()
      result(nil)

        // ✅ CONNECT: A correção principal está aqui
    case "connect":
      guard let deviceId = args["deviceId"] as? String, !deviceId.isEmpty else {
        result(FlutterError(code: "bad_args", message: "deviceId missing", details: nil))
        return
      }

      // 1. Aplica a configuração GLOBAL salva anteriormente (via .configure)
      // Isso garante que UUIDs do ESP32 sejam carregados antes de conectar.
      if transport == .ble {
        applyBleConfigFrom(bleConfig)
      } else if transport == .classic {
        applyClassicConfigFrom(classicConfig)
      } else {
        applyEpsonConfigFrom(epsonConfig)
      }

      // 2. Verifica se existe override específico nesta chamada (opcional)
      // Se não houver override, mantemos a config global.
      if transport == .ble {
        if let b = args["ble"] as? [String: Any] {
          applyBleConfigFrom(b)
        }
        // FIX: Removemos o 'else { applyBleConfigFrom(args) }'
        // Isso impedia que argumentos vazios resetassem o driver para Zebra.
      } else if transport == .classic {
        if let c = args["classic"] as? [String: Any] {
          applyClassicConfigFrom(c)
        }
      } else {
        if let e = args["epson"] as? [String: Any] {
          applyEpsonConfigFrom(e)
        }
      }

      manager.connect(deviceId: deviceId)
      result(nil)

    case "disconnect":
      manager.disconnect()
      result(nil)

    case "printRawBytes":
      guard let typed = args["data"] as? FlutterStandardTypedData else {
        result(FlutterError(code: "bad_args", message: "data missing", details: nil))
        return
      }
      manager.printRaw(data: typed.data)
      result(nil)

    case "getDiscoveredDevices":
      if transport == .epson {
        result(epson.snapshot())
      } else {
        result(store.snapshot(for: transport))
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Helpers: Apply BLE Config
  private func applyBleConfigFrom(_ args: [String: Any]) {
    // Extrai os valores do mapa (podem ser nulos, o Transport lida com defaults se necessário)
    let serviceUuid = args["serviceUuid"] as? String
    let writeCharacteristicUuid = args["writeCharacteristicUuid"] as? String

    let chunkSize = args["chunkSize"] as? Int

    // Dart manda ms, Swift transport usa us (microsegundos) internamente
    let chunkDelayMs = args["chunkDelayMs"] as? Int
    let chunkDelayUs: Int? = chunkDelayMs != nil ? max(0, chunkDelayMs!) * 1000 : nil

    let autoDisconnectMs = args["autoDisconnectMs"] as? Int

    ble.applyBleConfig(
        serviceUuid: serviceUuid,
        writeCharUuid: writeCharacteristicUuid,
        chunkSize: chunkSize,
        chunkDelayUs: chunkDelayUs,
        autoDisconnectMs: autoDisconnectMs
    )
  }

  // MARK: - Helpers: Apply Classic Config
  private func applyClassicConfigFrom(_ args: [String: Any]) {
    let preferredProtocol = args["preferredProtocol"] as? String
    let autoDisconnectMs = args["autoDisconnectMs"] as? Int

    classic.applyClassicConfig(
        preferredProtocol: preferredProtocol,
        autoDisconnectMs: autoDisconnectMs
    )
  }

  // MARK: - Helpers: Apply Epson ePOS Config
  private func applyEpsonConfigFrom(_ args: [String: Any]) {
    let portType = args["portType"] as? String
    let printerSeries = args["printerSeries"] as? String
    let modelLang = args["modelLang"] as? String
    let connectTimeoutMs = args["connectTimeoutMs"] as? Int
    let sendTimeoutMs = args["sendTimeoutMs"] as? Int
    let autoDisconnectMs = args["autoDisconnectMs"] as? Int

    epson.applyEpsonConfig(
        portType: portType,
        printerSeries: printerSeries,
        modelLang: modelLang,
        connectTimeoutMs: connectTimeoutMs,
        sendTimeoutMs: sendTimeoutMs,
        autoDisconnectMs: autoDisconnectMs
    )
  }
}

File: android/src/main/kotlin/br/com/bluethermal/blue_thermal_plus/BlueThermalPlusPlugin.kt
package br.com.bluethermal.blue_thermal_plus

import android.bluetooth.BluetoothManager
import android.content.Context

// ✅ CORREÇÃO 1: O import correto para NonNull no AndroidX
import androidx.annotation.NonNull

import br.com.bluethermal.blue_thermal_plus.core.TransportRouter
import br.com.bluethermal.blue_thermal_plus.shared.DeviceStore
import br.com.bluethermal.blue_thermal_plus.transports.ble.BleTransport
import br.com.bluethermal.blue_thermal_plus.transports.classic.ClassicTransport

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class BlueThermalPlusPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context

    private val store = DeviceStore()

    private val bleTransport by lazy {
        val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        BleTransport(context, manager.adapter, store)
    }

    private val classicTransport by lazy {
        val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        ClassicTransport(manager.adapter, store)
    }

    private val router by lazy {
        TransportRouter(bleTransport, classicTransport)
    }

    private var globalBleConfig: Map<String, Any>? = null
    private var globalClassicConfig: Map<String, Any>? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "blue_thermal_plus/methods")
        channel.setMethodCallHandler(this)
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "blue_thermal_plus/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        val sink: (Map<String, Any>) -> Unit = { payload -> events?.success(payload) }
        bleTransport.onEvent = sink
        classicTransport.onEvent = sink
    }

    override fun onCancel(arguments: Any?) {
        bleTransport.onEvent = null
        classicTransport.onEvent = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        val args = call.arguments as? Map<String, Any> ?: emptyMap()
        val transportStr = args["transport"] as? String

        // ✅ CORREÇÃO 2: Agora o router.manager aceita String? (nulo),
        // então o erro de Type Mismatch vai sumir.
        val manager = router.manager(transportStr)

        // Mantemos essa flag para a lógica de config
        val isBle = (transportStr == "ble" || transportStr == null)

        when (call.method) {
            "configure" -> {
                val b = args["ble"] as? Map<String, Any>
                if (b != null) {
                    globalBleConfig = b
                    bleTransport.applyConfig(b)
                }
                val c = args["classic"] as? Map<String, Any>
                if (c != null) {
                    globalClassicConfig = c
                    classicTransport.applyConfig(c)
                }
                result.success(null)
            }
            "startScan" -> {
                manager.startScan()
                result.success(null)
            }
            "stopScan" -> {
                manager.stopScan()
                result.success(null)
            }
            "connect" -> {
                val deviceId = args["deviceId"] as? String
                if (deviceId.isNullOrEmpty()) {
                    result.error("bad_args", "deviceId missing", null)
                    return
                }

                // Lógica de aplicar config antes de conectar
                if (isBle) {
                    bleTransport.applyConfig(globalBleConfig)
                } else {
                    classicTransport.applyConfig(globalClassicConfig)
                }

                // Override opcional se vier no connect
                if (isBle) {
                    val b = args["ble"] as? Map<String, Any>
                    if (b != null) bleTransport.applyConfig(b)
                } else {
                    val c = args["classic"] as? Map<String, Any>
                    if (c != null) classicTransport.applyConfig(c)
                }

                manager.connect(deviceId)
                result.success(null)
            }
            "disconnect" -> {
                manager.disconnect()
                result.success(null)
            }
            "printRawBytes" -> {
                val data = args["data"] as? ByteArray
                if (data == null) {
                    result.error("bad_args", "data missing", null)
                    return
                }
                manager.printRaw(data)
                result.success(null)
            }
            "getDiscoveredDevices" -> {
                val safeTransport = transportStr ?: "ble"
                result.success(store.snapshot(safeTransport))
            }
            else -> result.notImplemented()
        }
    }
}